tests/test_requests.py

```diff
@@ -315,6 +315,26 @@ def test_headers_on_session_with_None_are_not_sent(self, httpbin):
         prep = ses.prepare_request(req)
         assert 'Accept-Encoding' not in prep.headers
 
+    def test_headers_preserve_order(self, httpbin):
+        """Preserve order when headers provided as OrderedDict."""
+        ses = requests.Session()
+        ses.headers = OrderedDict()
```

Comment:
> they might use a list of tuples

In fact, they can't: https://github.com/kennethreitz/requests/issues/3038#issuecomment-193699603

`Session.headers` and `Request.headers` are required to be mappings: https://github.com/kennethreitz/requests/blob/master/requests/sessions.py#L55-L60